packages/vincent-tools/vincent-tool-sdk/src/lib/get-onchain-policy-params.ts

```diff
@@ -0,0 +1,97 @@
+import { abiDecodePolicyParameters } from "./abi-decode-policy-params";
+import { formatZodError } from "./format-zod-error";
+
+interface AllOnChainPolicyParams {
+    isPermitted: boolean;
+    appId: ethers.BigNumber;
+    appVersion: ethers.BigNumber;
+    policies: Policy[];
+}
+
+export const getOnChainPolicyParams = async (args: {
+    yellowstoneRpcUrl: string,
+    vincentContractAddress: string,
+    appDelegateeAddress: string,
+    agentWalletPkpTokenId: string,
+    toolIpfsCid: string,
+    policyIpfsCid: string,
+    policyUserParamsSchema?: z.infer<VincentPolicyDef['userParamsSchema']>
+}): Promise<z.infer<VincentPolicyDef['userParamsSchema'] | undefined>> => {
+    try {
+        const allOnChainPolicyParams = await _getAllOnChainPolicyParams(args);
+
+        // We exit early here because !allOnChainPolicyParams.isPermitted means appDelegateeAddress
+        // is not permitted to execute toolIpfsCid for the Vincent App on behalf of the agentWalletPkpTokenId
+        // and no further processing is needed
+        if (!allOnChainPolicyParams.isPermitted) {
+            throw new Error(`App Delegatee: ${args.appDelegateeAddress} is not permitted to execute Vincent Tool: ${args.toolIpfsCid} for App ID: ${allOnChainPolicyParams.appId.toString()} App Version: ${allOnChainPolicyParams.appVersion.toString()} using Agent Wallet PKP Token ID: ${args.agentWalletPkpTokenId}`);
+        }
+
+        const onChainPolicyParams = allOnChainPolicyParams.policies.find(
+            (policy) => policy.policyIpfsCid === args.policyIpfsCid
+        );
+
+        if (args.policyUserParamsSchema) {
+            if (!onChainPolicyParams) {
+                // All user Policy params can be optional, so if no parameters were set on-chain,
+                // we want to validate an empty object (i.e. no parameters) against the userParamsSchema
+                // to validate that there are no required Policy params
+                return parseOnChainPolicyParams(args.policyUserParamsSchema.parse({}));
+            } else {
+                const decodedPolicyParams = abiDecodePolicyParameters({ params: onChainPolicyParams.parameters });
+                return parseOnChainPolicyParams(args.policyUserParamsSchema.parse(decodedPolicyParams));
+            }
+        } else {
+            if (!onChainPolicyParams) {
+                return;
+            } else {
+                throw new Error(`Agent Wallet PKP Token ID: ${args.agentWalletPkpTokenId} has registered on-chain Policy parameters for Vincent App: ${allOnChainPolicyParams.appId.toString()} App Version: ${allOnChainPolicyParams.appVersion.toString()} Vincent Policy: ${args.policyIpfsCid} Vincent Tool: ${args.toolIpfsCid} but no userParamsSchema was defined by the Policy`);
+            }
+        }
+    } catch (error) {
+        throw new Error(`Error getting on-chain policy parameters (getOnChainPolicyParams): ${error instanceof Error ? error.message : String(error)}`);
+    }
```

Comment:
Given how explicit you made your errors for all of the child functions here, and with the changes i mentioned above, I think we can :hocho: this function-wide try/catch; it actually only serves to obfuscate the real error now that all of your child functions throw very explicit errors. :+1: Nice work